### ISSUE TITLE
DFOCC Gradients Not Symmetrized

### ISSUE THREAD

**Issue by JonathonMisiewicz** [psi4/psi4#1504]
Discovered while remedying test cases where reference numbers needed to be changed due to my upcoming #1143 fix. DFOCC gradients are not symmetrized. If you'd like an example, just look in the [reference output files](https://github.com/psi4/psi4/blob/master/tests/dfomp3-grad2/output.ref#L603-L604)! Symmetry of the overlap gradient is broken in the seventh decimal place. The other terms have some errors, but much smaller. That the overlap gradient is most afflicted is unsurprising, since the overlap integrals are contracted against one of the many "generalized Fock matrices". This one depends on the 1RDM and 2RDM, which are not currently symmetrized.

I'm hoping to include this in 1.3 with the #1143 fix, since this bug could lead to optimizations failing, as we saw with DFMP2 two weeks ago (#1483). Because individual terms are printed, the individual terms need to be symmetrized, and it may take me some time to find where the `symmetrize` calls need to be placed. If it takes too long, I'm fine with delaying to 1.4.

**Comment by CDSherrill** [comment]
I have been concerned for some time that there seems to be some little
symmetry-breaking noise in nearly all Psi4 gradients, and I have not taken
the time to follow up on it.  It puzzles me, because we never saw stuff
like this in Psi3... those gradients were already symmetric and did not
need a "symmetrize" function to be called.... The only part of Psi that
I've checked that seems immune to this problem is Daniel Crawford's CC
gradients, which are perfectly symmetric.

**Comment by andysim** [comment]
@CDSherrill - I don't think the noise in the gradients is a problem.  Psi3 routed all gradient computations through `cints`, which symmetrizes everything automagically in `src/bin/cints/Default_Deriv1/symmetrize_deriv1.cc`, which is called from `src/bin/cints/Default_Deriv1/deriv1.cc` 

@JonathonMisiewicz - the fix that I put in for DF-MP2 doesn't symmetrize the individual components of the gradient.  Only the total.  I didn't bother with the components just because they're only really for debugging and aren't printed by default.  I guess they probably should be symmetrized if they're printed out by a power user though.

**Comment by JonathonMisiewicz** [comment]
@andysim In DFOCC, the terms _are_ printed out by default, which is the main reason why I'm bothering. The individual terms were easier to find than I expected, so I already have all-but-one of the terms symmetrized.

**Comment by loriab** [comment]
`Matrix::symmetrize_gradient` does take a mol. could it be arranged to take the parent, not the clone?

**Comment by JonathonMisiewicz** [comment]
Not in a way I can see. The cloning is done Py-side, and all C sees is the clone. Once `dfocc` starts, control never returns to Python again until after the computation is complete.

**Comment by loriab** [comment]
so either DFOCC needs to take in the parent mol as well for symmetrization purposes or more of the driver portion of DFOCC needs to move py-side? I'm assuming these are some gradients that are intermediate in the DFOCC process, not the final result, which could be symmetrized at the end when control returns to `proc.py`.

**Comment by JonathonMisiewicz** [comment]
Correct, I need to symmetrize individual intermediates in the gradient expression that get printed.

Now that you mention it, I think moving more of the driver portion of DFOCC Py-side is the way to go. All I'd need to move is `dfgrad.cc` into `proc.py:run_dfocc_gradient`. `proc.py:run_dfocc_gradient` has access to the input molecule, so it can `symmetrize_gradient` correctly. Does this sound good? It may need to be its own PR, separate from the 1143 fix.
